### PR TITLE
Migrate to numpydoc docstrings

### PR DESCRIPTION
**Pull request by chaityacshah**
Migrate to numpydoc docstrings for several glue/core files (WIP for issue #466).

### CODE REVIEW

File: glue/core/command.py [hunk 1]
```diff
@@ -119,7 +122,9 @@ def do(self, cmd):
         """
         Execute and log a new command
 
-        :rtype: The return value of cmd.do()
+        Returns:
+
+            The return value of cmd.do()
```
Comment (astrofrog) [inline glue/core/command.py]: I think you can just remove the information about the return value here as it isn't interesting.

File: glue/core/command.py [hunk 2]
```diff
@@ -164,7 +175,10 @@ def can_undo_redo(self):
         """
         Return whether undo and redo options are possible
 
-        :rtype: (bool, bool) - Whether undo and redo are possible, respectively
+        Returns
+        -------
+        (bool, bool)
+            Whether undo and redo are possible, respectively
```
Comment (astrofrog) [inline glue/core/command.py]: I think the syntax here should be like:

```
Returns
--------
undo_possible : bool
    Whether undoing is possible
redo_possible : bool
    Whether redoing is possible
```

File: glue/core/command.py [hunk 3]
```diff
@@ -205,9 +219,12 @@ def undo(self, session):
 class NewDataViewer(Command):
     """Add a new data viewer to the application
 
-    :param viewer: The class of viewer to create
-    :param data: The data object to initialize the viewer with, or None
-    :type date: :class:`~glue.core.data.Data` or None
+    Parameters
+    ----------
+    viewer:
```
Comment (astrofrog) [inline glue/core/command.py]: Here and elsewhere where the type is a viewer, the type to include in the docstring should be ``glue.viewers.common.viewer.BaseViewer``

File: glue/core/data_factories/excel.py [hunk 4]
```diff
@@ -12,13 +12,21 @@
...
-    """ A factory for reading excel data using pandas.
-    :param path: path/to/file
-    :param sheet: The sheet to read. If `None`, all sheets are read.
-    :param kwargs: All other kwargs are passed to pandas.read_excel
-    :return: core.data.Data object.
+    """A factory for reading excel data using pandas.
+
+    Parameters
+    ----------
+    path : str`
```
Comment (astrofrog) [inline glue/core/data_factories/excel.py]: Remove the `

File: glue/core/data_factories/excel.py [hunk 5]
```diff
@@ -12,13 +12,21 @@
...
+        path/to/file
```
Comment (astrofrog) [inline glue/core/data_factories/excel.py]: Maybe write out 'Path to file'

File: glue/core/edit_subset_mode.py [hunk 6]
```diff
@@ -80,18 +80,18 @@ def update(self, d, new_state, focus_data=None, override_mode=None):
...
-
-        :param focus_data: The main data set in focus by the client,
-        if relevant. If a data set is in focus and has no subsets,
-        a new one will be created using new_state.
-
-        :param override_mode: Mode to use instead of EditSubsetMode.mode
-        :type override_mode: bool
+        Parameters
+        ----------
+        d : Data or DataCollection
```
Comment (astrofrog) [inline glue/core/edit_subset_mode.py]: Replace ``Data`` with

```
:class:`~glue.core.data.Data`
```

and similarly for ``DataCollection``

File: glue/core/edit_subset_mode.py [hunk 7]
```diff
@@ -80,18 +80,18 @@ def update(self, d, new_state, focus_data=None, override_mode=None):
...
+            Data or Collection to act upon
+        new_state : :class:`~glue.core.subset.SubsetState`
+            Subset state to combine with
+        focus_data :
```
Comment (astrofrog) [inline glue/core/edit_subset_mode.py]: The type here should be ``glue.core.data.Data``

File: glue/core/edit_subset_mode.py [hunk 8]
```diff
@@ -80,18 +80,18 @@ def update(self, d, new_state, focus_data=None, override_mode=None):
...
+            The main data set in focus by the client,
+            if relevant. If a data set is in focus and has no subsets,
+            a new one will be created using new_state.
+        override_mode : bool
```
Comment (astrofrog) [inline glue/core/edit_subset_mode.py]: The type here should be ``callable`` (I can explain later)

File: glue/core/message.py [hunk 9]
```diff
@@ -28,15 +28,23 @@ class Message(object):
     The message class family is hierarchical, and a client subscribing
     to a message class implicitly subscribes to all of its subclasses.
 
-    :attr sender: The object which sent the message
-    :attr tag: An optional string describing the message
+    Attributes
+    ----------
+    sender :
```
Comment (astrofrog) [inline glue/core/message.py]: When there is no specific type (as is the case here), just omit the ``:``.

File: glue/core/roi.py [hunk 10]
```diff
@@ -463,10 +487,10 @@ class VertexROIBase(Roi):
 
     def __init__(self, vx=None, vy=None):
         """
-        :param vx: initial x vertices
-        :type vx: list
-        :param vy: initial y vertices
-        :type vy: list
+        vx : list
+             initial x vertices
```
Comment (astrofrog) [inline glue/core/roi.py]: It looks like there are fives spaces in the indentation here but there should be four.

File: glue/core/message.py (hunk 9 above)
Comment (chaityacshah) [reply glue/core/message.py]: Is there a way to find out the type for a parameter that I am unaware of? Any `type` equivalent which can be used?

**Comment by chaityacshah** [comment]
@astrofrog- I have made a gist [here](https://gist.github.com/chaityacshah/4a5a3c35fe558099630f07c44727ddf9) which lists line numbers where I couldn't find either a data type or description. Can you please update this gist with the data types and descriptions so I can make changes accordingly?

Thanks!

**Comment by chaityacshah** [comment]
Is there a particular pattern that needs to be followed when making commits? For e.g., should there be a minimum number of files where changes are made before a 'commit'? Should I continue pushing to the same `master` branch for this pull request?

**Comment by astrofrog** [comment]
No rules for when to commit - up to you :) Yes just keep adding commits to the master branch. In future it would be better to use other branches than master so that you can have multiple PRs open, but this will be fine for now.

**Comment by astrofrog** [comment]
@chaityacshah - sorry for the delay. Here is some info:

https://github.com/glue-viz/glue/blob/master/glue/core/state.py#L179 -> I think key doesn't have a fixed type, it can be any Python object. So just don't give the type in the docstring.

https://github.com/glue-viz/glue/blob/master/glue/core/subset_group.py#L47 -> description should be 'The subset group this subset belongs to'

https://github.com/glue-viz/glue/blob/master/glue/core/visual.py#L17 -> the types should be:

    :param color: str
    :param alpha: float
    :param linewidth:float or int
    :param linestyle: str, one of (``'solid' | 'dashed' | 'dash-dot' | 'dotted' | 'none'``)
    :param marker: str (``'o' | 's' | '^' | etc``)
    :param markersize: int

Let me know if you have other questions!

**Comment by astrofrog** [comment]
I am going to close this issue due to lack of activity - but @chaityacshah feel free to open a new pull request if you have time to work on this again in future!

**Pull request closed without merging**